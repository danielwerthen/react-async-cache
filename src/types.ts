import { Observable } from 'rxjs';

export type AsyncResult<T, ERR> = {
  pending: boolean;
  result?: T;
  error?: ERR;
  refresh: () => Promise<AsyncResult<T, ERR>>;
};

export type ObserveValue = <T>(input: Observable<T>) => Promise<T>;

export type AsyncFactory<T> = (observe: ObserveValue) => Promise<T>;
