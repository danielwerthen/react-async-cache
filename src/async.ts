import {
  BehaviorSubject,
  combineLatest,
  from,
  isObservable,
  Observable,
  of,
  ReplaySubject,
  Subject,
  Subscription,
} from 'rxjs';
import {
  catchError,
  concatMap,
  debounceTime,
  finalize,
  map,
  publishReplay,
  refCount,
  skip,
  startWith,
  switchMap,
  take,
  tap,
} from 'rxjs/operators';
import { exhaustMapWithTrailing, useObservedProp } from './utils';
import { AsyncFactory, AsyncResult, SharedAsync } from './types';
import { DependencyList, useCallback, useMemo } from 'react';
import { useAsyncBase } from './useAsyncBase';

export function observeAsync<T, ERR = unknown>(
  factories: Observable<AsyncFactory<T>>,
  refreshSubject: Subject<unknown> = new Subject()
): BehaviorSubject<AsyncResult<T, ERR>> {
  const subj: BehaviorSubject<AsyncResult<T, ERR>> = new BehaviorSubject<
    AsyncResult<T, ERR>
  >({
    pending: true,
    refresh: () => {
      refreshSubject.next(0);
      return subj.pipe(skip(1), take(1)).toPromise();
    },
  });
  const obsMap = new WeakMap<Observable<unknown>, ReplaySubject<unknown>>();
  const subMap = new WeakMap<Observable<unknown>, Subscription>();
  const resolveDependant = <T>(observable: Observable<T>): Promise<T> => {
    if (!obsMap.has(observable)) {
      const published = new ReplaySubject(1);
      obsMap.set(observable, published);
      subMap.set(
        observable,
        observable
          .pipe(
            map((item, idx) => {
              if (idx !== 0) {
                refreshSubject.next(0);
              }
              return item;
            })
          )
          .subscribe(published)
      );
    }
    return (obsMap.get(observable) as Observable<T>).pipe(take(1)).toPromise();
  };
  const unsubscribeDependent = (observable: Observable<unknown>) => {
    const obs = obsMap.get(observable);
    const sub = subMap.get(observable);
    if (obs) {
      obsMap.delete(observable);
    }
    if (sub) {
      sub.unsubscribe();
      subMap.delete(observable);
    }
  };
  let prevDependants: Observable<unknown>[] = [];
  let subscription: Subscription;
  const obs = combineLatest([
    factories,
    refreshSubject.pipe(debounceTime(0), startWith(0)),
  ]).pipe(
    exhaustMapWithTrailing(([factory], index) => {
      const dependants: Observable<unknown>[] = [];
      const promise = factory(observable => {
        const obs = resolveDependant(observable);
        dependants.push(observable);
        return obs;
      });
      return (isObservable(promise)
        ? promise
        : from(promise).pipe(
            concatMap(result => {
              if (isObservable(result)) {
                return result;
              }
              return of(result);
            })
          )
      ).pipe(
        finalize(() => {
          prevDependants.forEach(dep => {
            if (!dependants.includes(dep)) {
              unsubscribeDependent(dep);
            }
          });
          prevDependants = dependants;
        }),
        map(result => ({
          pending: false,
          result,
          refresh: subj.value.refresh,
        })),
        catchError(error =>
          of({
            pending: false,
            error: error as ERR,
            refresh: subj.value.refresh,
          })
        ),
        index > 0
          ? startWith({
              pending: true,
              result: subj.value.result,
              refresh: subj.value.refresh,
            })
          : tap(() => void 0)
      );
    }),
    finalize(() => {
      prevDependants.forEach(dep => {
        unsubscribeDependent(dep);
      });
      if (subscription) {
        subscription.unsubscribe();
        refreshSubject.complete();
      }
    })
  );
  subscription = obs.subscribe(subj);
  const oldComplete = subj.complete;
  return Object.assign(subj, {
    complete() {
      oldComplete.apply(subj);
      subscription.unsubscribe();
    },
  });
}

/**
 * `useAsync` will only recompute the async result when one of the `dependencies` has changed.
 * The factory function gets an observe function as its first parameter. This function can turn an
 * observable into a promise, and update the async result once the observable emits a new value.
 * @param factory Async factory method which will be provided with the observe function.
 * @param dependencies
 */
export function useAsync<T, ERR>(
  factory: AsyncFactory<T>,
  dependencies: DependencyList
): AsyncResult<T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  const subject = useMemo(() => {
    return observeAsync<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  return useAsyncBase(subject, subject.value);
}

/**
 * Compared to `useAsync`, this will allow the resulting value be shared by any components.
 * The resulting observable is published and refcounted to ensure the same async result is shared between subscribers.
 * The factory will be trigged when the observable is resubscribed after a period of no subscriptions.
 * @param factory Async factory method which will be provided with the observe function.
 * @param finalizeFn The resulting observable will call this callback whenever there is no subscriptions ongoing.
 */
export function shareAsync<OUTPUT, ERR = unknown>(
  factory: AsyncFactory<OUTPUT>,
  finalizeFn: () => void = () => void 0
): SharedAsync<OUTPUT, ERR> {
  const obs = of(0).pipe(
    switchMap(() => {
      return observeAsync<OUTPUT, ERR>(of(factory));
    }),
    skip(1),
    finalize(finalizeFn),
    publishReplay(1),
    refCount()
  );
  const refresh = () => {
    return obs
      .pipe(take(1))
      .toPromise()
      .then(item => item.refresh());
  };
  return Object.assign(obs.pipe(startWith({ pending: true, refresh })), {
    useSubscribe() {
      return useAsyncBase(obs, { pending: true, refresh });
    },
    refresh,
  });
}

/**
 * Sometimes you want to share a set of similar async observables, cached by a `string`. Otherwise same as `shareAsync`.
 * @param options
 */
export function sharedAsyncMap<INPUT, OUTPUT, ERR = unknown>(
  factory: (input: INPUT) => AsyncFactory<OUTPUT>,
  getKey: (input: INPUT) => string
): (input: INPUT) => SharedAsync<OUTPUT, ERR> {
  const store: {
    [key: string]: SharedAsync<OUTPUT, ERR>;
  } = {};
  const getResource = (input: INPUT) => {
    const key = getKey(input);
    if (!store[key]) {
      const res = shareAsync<OUTPUT, ERR>(factory(input), () => {
        delete store[key];
      });
      store[key] = res;
    }
    return store[key];
  };
  return getResource;
}
