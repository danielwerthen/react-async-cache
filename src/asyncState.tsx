import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
} from 'react';
import { BehaviorSubject, from, isObservable, Subject } from 'rxjs';
import {
  concatMap,
  distinctUntilChanged,
  finalize,
  map,
  tap,
} from 'rxjs/operators';
import {
  AsyncReducer,
  AsyncState,
  AsyncAction,
  AsyncStateContext,
} from './types';
import { useInitialize, useSubscribe } from './utils';

/**
 * `asyncState` allows you to share redux-like state without the boilerplate. It utilies a queue to ensure that any dispatched action
 * is reflect in the shared state in a sequential way. If two actions are dispatched at the same time, the first one will be exhausted before the next action is updating the state.
 * The dispatch function returns a promise resolves once the entire dispatched action has been exhausted.
 * An action can be a single state value, or a function that returns a single, promise or observable of state values.
 * @param initialValue
 * @param reducer
 */
export function observeAsyncState<STATE, ACTION = STATE>(
  initialValue: STATE,
  reducer: AsyncReducer<STATE, ACTION>
): AsyncState<STATE, ACTION> {
  const queue = new Subject<[AsyncAction<STATE, ACTION>, (r: STATE) => void]>();
  const state$ = new BehaviorSubject<STATE>(initialValue);
  const sub = queue
    .pipe(
      concatMap(([applicator, resolver]) => {
        const action =
          applicator instanceof Function
            ? applicator(state$.value)
            : applicator;
        let lastValue: STATE;
        const observableAction = isObservable(action)
          ? action
          : from(Promise.resolve(action));
        return observableAction.pipe(
          concatMap(action => {
            const next = reducer(state$.value, action);
            return isObservable(next) ? next : Promise.resolve(next);
          }),
          tap(res => (lastValue = res)),
          finalize(() => resolver(lastValue))
        );
      })
    )
    .subscribe(state$);
  return Object.assign(state$.asObservable(), {
    dispatch(action: AsyncAction<STATE, ACTION>): Promise<STATE> {
      return new Promise<STATE>(resolve => {
        queue.next([action, resolve]);
      });
    },
    getValue(): STATE {
      return state$.value;
    },
    useSubscribe(): STATE {
      return useSubscribe(state$, state$.value);
    },
    useSelect<OUTPUT>(
      selector: (state: STATE) => OUTPUT,
      deps: unknown[]
    ): OUTPUT {
      const callback = useCallback(selector, deps);
      const selected = useMemo(
        () => state$.pipe(map(callback), distinctUntilChanged()),
        [callback]
      );
      const initial = useInitialize(() => selector(state$.value));
      return useSubscribe(selected, initial);
    },
    unsubscribe() {
      sub.unsubscribe();
      state$.complete();
      queue.unsubscribe();
    },
  });
}

export const shareAsyncState = observeAsyncState;

export function useAsyncState<STATE, ACTION = STATE>(
  initialValue: STATE,
  reducer: AsyncReducer<STATE, ACTION>,
  deps: unknown[]
) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const state$ = useMemo(() => observeAsyncState(initialValue, reducer), [
    reducer,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    ...deps,
  ]);
  const state = state$.useSubscribe();
  return [state, state$.dispatch];
}

/**
 * Sometime you want encapsulate the `asyncState` in a context, to allow for concurrent async states across an application.
 * @param stateFactory
 */
export function asyncStateContext<STATE, ACTION>(
  stateFactory: (initial?: STATE) => AsyncState<STATE, ACTION>
): AsyncStateContext<STATE, ACTION> {
  const context = createContext(stateFactory());
  const Provider: React.FC<{ initialState?: STATE }> = ({
    children,
    initialState,
  }) => {
    const ctx = useInitialize(() => stateFactory(initialState));
    useEffect(() => () => ctx.unsubscribe(), [ctx]);
    return <context.Provider value={ctx}>{children}</context.Provider>;
  };
  return {
    Provider,
    useSubscribe: () => {
      const state = useContext(context);
      return state.useSubscribe();
    },
    useSelect<OUTPUT>(
      selector: (state: STATE) => OUTPUT,
      deps: unknown[]
    ): OUTPUT {
      const state = useContext(context);
      return state.useSelect(selector, deps);
    },
    useDispatch: () => {
      const state = useContext(context);
      const dispatchRef = useRef(state.dispatch);
      dispatchRef.current = state.dispatch;
      return useCallback((...args) => dispatchRef.current(...args), [
        dispatchRef,
      ]);
    },
  };
}
