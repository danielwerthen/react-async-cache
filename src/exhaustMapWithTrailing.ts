import {
  from,
  Observable,
  ObservableInput,
  OperatorFunction,
  Subject,
} from 'rxjs';
import { exhaustMap, finalize, throttle } from 'rxjs/operators';

/**
 * Credit to: https://github.com/ReactiveX/rxjs/issues/5004
 * @param project
 */
export function exhaustMapWithTrailing<T, R>(
  project: (value: T, index: number) => ObservableInput<R>
): OperatorFunction<T, R> {
  return (source): Observable<R> => {
    const release = new Subject();

    return source.pipe(
      throttle(() => release, {
        leading: true,
        trailing: true,
      }),
      // TODO: Upgrade to TypeScript 3.6.3 when available and remove the cast
      // https://github.com/microsoft/TypeScript/issues/33131
      exhaustMap((value, index) =>
        from(project(value, index)).pipe(
          finalize(() => {
            release.next();
          })
        )
      ) as OperatorFunction<T, R>
    );
  };
}
