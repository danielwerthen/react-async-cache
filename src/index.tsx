import {
  catchError,
  concatMap,
  finalize,
  map,
  publishReplay,
  refCount,
  skip,
  startWith,
  switchMap,
  take,
  tap,
} from 'rxjs/operators';
import { BehaviorSubject, from, Observable, of, Subject } from 'rxjs';
import { fromFetch as defaultFromFetch } from 'rxjs/fetch';
import {
  DependencyList,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { AsyncFactory, AsyncResult, AsyncBase, AsyncCallback } from './types';
import { observeAsync } from './observeAsync';

class Monitor {
  usingPending = false;
  usingError = false;

  wrap<AR extends AsyncBase<unknown, unknown>>(result: AR): AR {
    if (this.usingPending && this.usingError) {
      return result;
    }
    return new Proxy(result, {
      get: (target, prop, receiver) => {
        switch (prop) {
          case 'pending':
            this.usingPending = true;
            break;
          case 'error':
            this.usingError = true;
            break;
        }
        return Reflect.get(target, prop, receiver);
      },
    });
  }
}

export function useSubscribe<T>(observable: Observable<T>, initialValue: T): T {
  const [state, setState] = useState(initialValue);
  useEffect(() => {
    const sub = observable.subscribe(setState);
    return () => sub.unsubscribe();
  }, [observable]);
  return state;
}

export function useAsyncBase<S extends AsyncBase<unknown, unknown>>(
  result$: Observable<S>,
  initialValue: S
): S {
  const monitor = useMemo(() => new Monitor(), []);
  const [output, setOutput] = useState<S>(initialValue);
  const outputRef = useRef<S>(output);
  outputRef.current = output;
  useEffect(() => {
    const sub = result$.subscribe(item => {
      if (item.pending) {
        if (monitor.usingPending) {
          setOutput(item);
        }
      } else {
        const old = outputRef.current;
        if (
          old.result !== item.result ||
          old.error !== item.error ||
          old.pending !== item.pending
        ) {
          setOutput(item);
        }
      }
    });
    return () => sub.unsubscribe();
  }, [result$, monitor, outputRef]);
  const final = useMemo(
    () => monitor.wrap(output),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [monitor, output]
  );
  if (!monitor.usingError) {
    if (output.error !== undefined) {
      throw output.error;
    }
  }
  return final;
}

/**
 * `useAsync` will only recompute the async result when one of the `dependencies` has changed.
 * The factory function gets an observe function as its first parameter. This function can turn an
 * observable into a promise, and update the async result once the observable emits a new value.
 * @param factory
 * @param dependencies
 */
export function useAsync<T, ERR>(
  factory: AsyncFactory<T>,
  dependencies: DependencyList
): AsyncResult<T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  const subject = useMemo(() => {
    return observeAsync<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  return useAsyncBase(subject, subject.value);
}

export type SharedAsync<OUTPUT, ERR = unknown> = Observable<
  AsyncResult<OUTPUT, ERR>
> & {
  useSubscribe(): AsyncResult<OUTPUT, ERR>;
  refresh(): Promise<AsyncResult<OUTPUT, ERR>>;
};

export function shareAsync<OUTPUT, ERR = unknown>(
  factory: AsyncFactory<OUTPUT>,
  finalizeFn: () => void = () => void 0
): SharedAsync<OUTPUT, ERR> {
  const obs = of(0).pipe(
    switchMap(() => {
      return observeAsync<OUTPUT, ERR>(of(factory));
    }),
    skip(1),
    finalize(finalizeFn),
    publishReplay(1),
    refCount()
  );
  const refresh = () => {
    return obs
      .pipe(take(1))
      .toPromise()
      .then(item => item.refresh());
  };
  return Object.assign(obs.pipe(startWith({ pending: true, refresh })), {
    useSubscribe() {
      return useSubscribe(obs, { pending: true, refresh });
    },
    refresh,
  });
}

export function sharedAsyncFactory<INPUT, OUTPUT, ERR = unknown>({
  factory,
  getKey,
}: {
  factory: (input: INPUT) => AsyncFactory<OUTPUT>;
  getKey: (input: INPUT) => string;
}): (input: INPUT) => SharedAsync<OUTPUT, ERR> {
  const store: {
    [key: string]: SharedAsync<OUTPUT, ERR>;
  } = {};
  const getResource = (input: INPUT) => {
    const key = getKey(input);
    if (!store[key]) {
      const res = shareAsync<OUTPUT, ERR>(factory(input), () => {
        delete store[key];
      });
      store[key] = res;
    }
    return store[key];
  };
  return getResource;
}

export function useObservedProp<A>(a: A): Observable<A> {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const subject = useMemo(() => new BehaviorSubject<A>(a), []);
  useEffect(() => {
    if (subject.value !== a) {
      subject.next(a);
    }
  }, [a, subject]);
  useEffect(() => () => subject.complete(), [subject]);
  return useMemo(() => subject.asObservable(), [subject]);
}

export function createSharedFetch<T, INPUT extends unknown[], ERR = unknown>(
  init$: Observable<RequestInit>,
  request: (...input: INPUT) => string,
  selector: (response: Response) => Promise<T>,
  fromFetch: typeof defaultFromFetch = defaultFromFetch
): (...input: INPUT) => SharedAsync<T, ERR> {
  const shared = sharedAsyncFactory<string, T, ERR>({
    factory(input: string) {
      return async observe => {
        const init = await observe(init$);
        return observe(
          fromFetch(input, {
            ...init,
            selector,
          })
        );
      };
    },
    getKey(input: string) {
      return input;
    },
  });
  return (...input: INPUT) => {
    return shared(request(...input));
  };
}

function isPromise<T>(obj: any): obj is Promise<T> {
  return obj && typeof obj.then === 'function';
}

function isObservable<T>(obj: any): obj is Observable<T> {
  return obj && typeof obj.subscribe === 'function';
}

export type AsyncState<T> = Observable<T> & {
  dispatch: (
    action: T | ((v: T) => T | Promise<T> | Observable<T>)
  ) => Promise<T>;
  getValue: () => T;
  useSubscribe: () => T;
  unsubscribe: () => void;
};

export function asyncState<T>(initialValue: T): AsyncState<T> {
  const queue = new Subject<
    [T | ((v: T) => T | Promise<T> | Observable<T>), (r: T) => void]
  >();
  const state$ = new BehaviorSubject<T>(initialValue);
  const sub = queue
    .pipe(
      concatMap(([applicator, resolver]) => {
        if (applicator instanceof Function) {
          const next = applicator(state$.value);
          if (isObservable(next)) {
            let lastValue: T;
            return next.pipe(
              tap(res => (lastValue = res)),
              finalize(() => resolver(lastValue))
            );
          } else if (isPromise(next)) {
            return from(
              next.then(pass => {
                resolver(pass);
                return pass;
              })
            );
          }
          resolver(next);
          return of(next);
        }
        resolver(applicator);
        return of(applicator);
      })
    )
    .subscribe(state$);
  return Object.assign(state$.asObservable(), {
    dispatch(
      action: T | ((v: T) => T | Promise<T> | Observable<T>)
    ): Promise<T> {
      return new Promise<T>(resolve => {
        queue.next([action, resolve]);
      });
    },
    getValue(): T {
      return state$.value;
    },
    useSubscribe(): T {
      return useSubscribe(state$, state$.value);
    },
    unsubscribe() {
      sub.unsubscribe();
      state$.complete();
    },
  });
}

export function asyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factories: Observable<(...input: INPUT) => Promise<T>>
) {
  const execute: (
    ...inputs: INPUT
  ) => Promise<AsyncCallback<INPUT, T, ERR>> = async (...input: INPUT) => {
    return await state$.dispatch(() => {
      return factories.pipe(
        take(1),
        switchMap(fn => fn(...input)),
        map(result => ({ pending: false, result, execute })),
        catchError(err => of({ pending: false, error: err as ERR, execute })),
        startWith({
          pending: true,
          execute,
        })
      );
    });
  };
  const state$ = asyncState<AsyncCallback<INPUT, T, ERR>>({
    pending: false,
    execute,
  });
  return state$;
}

export function useAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factory: (...input: INPUT) => Promise<T>,
  dependencies: unknown[]
): AsyncCallback<INPUT, T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const state$ = useMemo(() => asyncCallback<INPUT, T, ERR>(factories), [
    factories,
  ]);
  return useAsyncBase(state$, state$.getValue());
}
