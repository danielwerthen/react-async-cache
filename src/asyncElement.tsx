import React, {
  DependencyList,
  ReactElement,
  ReactNode,
  useCallback,
  useMemo,
} from 'react';
import { observeAsync } from './async';
import { AsyncFactory } from './types';
import { useAsyncBase } from './useAsyncBase';
import { useObservedProp } from './utils';

export function useAsyncElement<T extends ReactElement<any, any> | null, ERR>(
  factory: AsyncFactory<T>,
  dependencies: DependencyList
): ReactNode {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  const subject = useMemo(() => {
    return observeAsync<T, ERR>(factories);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [factories]);
  const Comp = useCallback(() => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const { result } = useAsyncBase(subject, subject.value);
    return result || null;
  }, [subject]);

  return <Comp />;
}
