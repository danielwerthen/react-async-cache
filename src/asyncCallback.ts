import { useCallback, useMemo } from 'react';
import { Observable, of } from 'rxjs';
import { catchError, map, startWith, switchMap, take } from 'rxjs/operators';
import { observeAsyncState } from './asyncState';
import { AsyncCallback } from './types';
import { useAsyncBase } from './useAsyncBase';
import { useObservedProp } from './utils';

export function observeAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factories: Observable<(...input: INPUT) => Promise<T>>
) {
  const execute: (
    ...inputs: INPUT
  ) => Promise<AsyncCallback<INPUT, T, ERR>> = async (...input: INPUT) => {
    return await state$.dispatch(() => {
      return factories.pipe(
        take(1),
        switchMap(fn => fn(...input)),
        map(result => ({ pending: false, result, execute })),
        catchError(err => of({ pending: false, error: err as ERR, execute })),
        startWith({
          pending: true,
          execute,
        })
      );
    });
  };
  const state$ = observeAsyncState<AsyncCallback<INPUT, T, ERR>>(
    {
      pending: false,
      execute,
    },
    (_state, action) => action
  );
  return Object.assign(state$, {
    execute(...input: INPUT) {
      return state$
        .pipe(
          take(1),
          switchMap(item => item.execute(...input))
        )
        .toPromise();
    },
    useSubscribe() {
      return useAsyncBase(state$, state$.getValue());
    },
    useSelect<O>(
      selector: (state: AsyncCallback<INPUT, T, ERR>) => O,
      deps: unknown[]
    ) {
      return state$.useSelect(selector, deps);
    },
  });
}

export function shareAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factory: (...input: INPUT) => Promise<T>
) {
  return observeAsyncCallback<INPUT, T, ERR>(of(factory));
}

/**
 * Use this instead of `useAsync` if you want to control the timing of the asynchronous operations. Operations will be executed in an enforced sequence, using an underlying `asyncState`.
 * @param factory
 * @param dependencies
 */
export function useAsyncCallback<INPUT extends unknown[], T, ERR = unknown>(
  factory: (...input: INPUT) => Promise<T>,
  dependencies: unknown[]
): AsyncCallback<INPUT, T, ERR> {
  const callback = useCallback(factory, dependencies);
  const factories = useObservedProp(callback);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const state$ = useMemo(() => observeAsyncCallback<INPUT, T, ERR>(factories), [
    factories,
  ]);
  return useAsyncBase(state$, state$.getValue());
}
