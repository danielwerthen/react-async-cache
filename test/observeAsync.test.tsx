import { BehaviorSubject, interval, of } from 'rxjs';
import { AsyncResult } from '../src/types';
import { observeAsync } from '../src/async';
import { filter, map, take, toArray } from 'rxjs/operators';
import { shareAsync, observeAsyncState, observeAsyncCallback } from '../src';

function sleep(ms: number) {
  return new Promise(res => setTimeout(res, ms));
}

let WRef: any = (global as any).WeakRef;

let subjects: any[] = [];

function register(obj: any) {
  if (!subjects) {
    subjects = [];
  }
  if (!WRef) {
    WRef = (global as any).WeakRef;
  }
  subjects.push(new WRef(obj));
}

jest.mock('rxjs/internal/Subscription', () => {
  const actual = jest.requireActual('rxjs/internal/Subscription');
  class Subscription extends actual.Subscription {
    constructor(...args: []) {
      super(...args);
      register(this);
    }
  }
  return { ...actual, Subscription };
});

jest.mock('rxjs/internal/Observable', () => {
  const actual = jest.requireActual('rxjs/internal/Observable');
  class Observable extends actual.Observable {
    constructor(...args: []) {
      super(...args);
      register(this);
    }
  }
  return { ...actual, Observable };
});

jest.mock('rxjs', () => {
  const actual = jest.requireActual('rxjs');
  class Subject extends actual.Subject {
    constructor() {
      super();
      register(this);
    }
  }

  class BehaviorSubject extends actual.BehaviorSubject {
    constructor(...args: any[]) {
      super(...args);
      register(this);
    }
  }

  class ReplaySubject extends actual.ReplaySubject {
    constructor(...args: any[]) {
      super(...args);
      register(this);
    }
  }

  class Observable extends actual.Observable {
    constructor(...args: any[]) {
      super(...args);
      register(this);
    }
  }
  return {
    ...actual,
    Subject,
    BehaviorSubject,
    ReplaySubject,
    Observable,
  };
});

function verify(fn: () => Promise<void>) {
  return async () => {
    subjects.splice(0, subjects.length);
    global.gc();
    await fn();
    await sleep(10);
    global.gc();
    await sleep(10);
    const refs = subjects.map(sub => sub.deref());
    const notClean = refs.some(v => v);
    if (notClean) {
      throw new Error('Observables was not garbage collected in time');
    }
  };
}

describe('ObserveAsync', () => {
  it(
    'works with simple case',
    verify(async () => {
      const observed: BehaviorSubject<AsyncResult<any, unknown>> = observeAsync(
        of(async () => 4)
      );
      const final = observed.pipe(take(2), toArray()).toPromise();
      await sleep(0);
      expect(await final).toMatchSnapshot();
      await sleep(100);
    })
  );

  it(
    'works with some dependants',
    verify(async () => {
      const dependantA = new BehaviorSubject(7);
      const dependantB = new BehaviorSubject(17);
      const timer = interval(500);
      const observed: BehaviorSubject<AsyncResult<any, unknown>> = observeAsync(
        of(async observe => {
          const depA = await observe(dependantA);
          const depB = await observe(dependantB);
          observe(timer);
          return [depA, depB];
        })
      );
      const final = observed
        .pipe(
          filter(item => !item.pending),
          map(item => item.result),
          toArray()
        )
        .toPromise();
      dependantA.next(6);
      await sleep(0);
      dependantA.next(7);
      await sleep(0);
      dependantA.next(5);
      dependantB.next(8);
      dependantB.next(8);
      dependantB.next(3);
      dependantB.next(2);
      await sleep(0);
      dependantA.next(9);
      await sleep(0);
      observed.complete();
      dependantA.complete();
      dependantB.complete();
      expect(await final).toMatchSnapshot();
    })
  );
});

describe('Share async', () => {
  it(
    'should unsubscribe as expected',
    verify(async () => {
      let finals = 0;
      let starts = 0;
      const shared = shareAsync(
        async () => {
          starts += 1;
          return 5;
        },
        () => {
          finals += 1;
        }
      );
      const sub1 = shared.subscribe();
      const sub2 = shared.subscribe();
      const sub3 = shared.subscribe();
      sub1.unsubscribe();
      sub2.unsubscribe();
      sub3.unsubscribe();
      expect(finals).toEqual(1);
      expect(starts).toEqual(1);
      const sub4 = shared.subscribe();
      const sub5 = shared.subscribe();
      expect(starts).toEqual(2);
      expect(finals).toEqual(1);
      sub4.unsubscribe();
      sub5.unsubscribe();
      expect(finals).toEqual(2);
      expect(starts).toEqual(2);
    })
  );
});

describe('asyncState', () => {
  it(
    'should work',
    verify(async () => {
      const state = observeAsyncState(15, (_state, action) => action);
      const promise = state.pipe(toArray()).toPromise();
      state.dispatch(13);
      state.dispatch(() => 23);
      state.dispatch(v => v + 23);
      await sleep(1);
      state.unsubscribe();
      expect(await promise).toMatchSnapshot();
    })
  );
});

describe('asyncCallback', () => {
  it(
    'should produce the expected array',
    verify(async () => {
      const state = observeAsyncCallback(
        of(async (item: string) => {
          return `Modified(${item})`;
        })
      );
      const promise = state.pipe(toArray()).toPromise();
      await state.execute('test');
      await state.execute('test2');
      await state.execute('test3');
      state.unsubscribe();
      expect(await promise).toMatchSnapshot();
    })
  );
});
