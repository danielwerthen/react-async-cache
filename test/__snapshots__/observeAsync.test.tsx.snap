// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObserveAsync works with simple case 1`] = `
Array [
  Object {
    "pending": true,
    "refresh": [Function],
  },
  Object {
    "pending": false,
    "refresh": [Function],
    "result": 4,
  },
]
`;

exports[`ObserveAsync works with some dependants 1`] = `
Array [
  Array [
    7,
    17,
  ],
  Array [
    6,
    17,
  ],
  Array [
    7,
    17,
  ],
  Array [
    5,
    2,
  ],
  Array [
    9,
    2,
  ],
]
`;

exports[`asyncState should work 1`] = `
Array [
  15,
  13,
  23,
  46,
]
`;
